presto-main/src/main/java/io/prestosql/sql/planner/SymbolsExtractor.java

```diff
@@ -46,6 +49,7 @@ private SymbolsExtractor() {}
     {
         ImmutableSet.Builder<Symbol> uniqueSymbols = ImmutableSet.builder();
         extractExpressions(node).forEach(expression -> uniqueSymbols.addAll(extractUnique(expression)));
+        uniqueSymbols.addAll(extractOrderBySymbols(node, true, noLookup()));
```

Comment:
So far correlated symbols were only allowed in expressions, if we change this then we would need change the logic in some other places as well possibly.

I think it would be better to inject correlation into a projection and then use projected symbol in order by. 

Was correlated symbol inlined by unalias optimizer?

Reply:
Actually a correlated symbol might appear in an `OrderingScheme`.
`PlanNodeDecorrelator.containsCorrelation()` does't check that -- and if the query fails with `correlated subquery not supported`, this is due to another reason.
This symbol extractor is intended to tighten up the search for correlated symbols, so that in case when we are able to decorrelate more nodes (e.g. ProjectNode), the query doesn't fail because some correlated symbol slipped through.

Reply:
> I think it would be better to inject correlation into a projection and then use projected symbol in order by.

I agree with @kokosing that this is the way to go. However, it seems like a much bigger effort and `SymbolsExtractor` should collect all symbols anyway.

Reply:
> So far correlated symbols were only allowed in expressions, [..]

This is not an assumption I would like to make.
The planner should be allowed to use correlated symbols whenever it needs them, allowing it to be as simple as possible.
In particular it should not do any translation based on equality predicates of the subquery. This is a job for the optimizer.

Reply:
> The planner should be allowed to use correlated symbols whenever it needs them, allowing it to be as simple as possible.

That makes the plan and optimizers more complicated and less expressive3. , because:
1. Rules and optimizations need to understand that some symbols might appear out of nowhere
2. When creating plan nodes, we cannot validate symbol origins.
3. We cannot create plans for certain queries, e.g.
```
SELECT b, (SELECT b FROM ...) FROM (VALUES ...) t(b)
```
(plan does not distinguish between `b` from subquery and outer relation.